plugins/ccgx/fu-ccgx-pure-hid-device.c

```diff
@@ -0,0 +1,472 @@
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "image VID:PID mismatch, expected %04X:%04X, got %04X:%04X",
+			    fu_usb_device_get_vid(FU_USB_DEVICE(self)),
+			    fu_usb_device_get_pid(FU_USB_DEVICE(self)),
+			    vid,
+			    pid);
+		return NULL;
+	}
+
+	return g_steal_pointer(&firmware);
+}
+
+static gboolean
+fu_ccgx_pure_hid_write_row(FuCcgxPureHidDevice *self,
+			   guint16 address,
+			   const guint8 *row,
+			   gsize row_len,
+			   GError **error)
+{
+	g_autoptr(GByteArray) st_hdr = fu_struct_ccgx_pure_hid_write_hdr_new();
+
+	fu_struct_ccgx_pure_hid_write_hdr_set_pd_resp(st_hdr,
+						      FU_CCGX_PD_RESP_FLASH_READ_WRITE_CMD_SIG);
+	fu_struct_ccgx_pure_hid_write_hdr_set_addr(st_hdr, address);
+	if (!fu_memcpy_safe(st_hdr->data,
+			    st_hdr->len,
+			    FU_STRUCT_CCGX_PURE_HID_WRITE_HDR_OFFSET_DATA,
+			    row,
+			    row_len,
+			    0,
+			    self->flash_row_size,
+			    error))
+		return FALSE;
+
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(self),
+				      st_hdr->data[0],
+				      st_hdr->data,
+				      st_hdr->len,
+				      FU_CCGX_PURE_HID_DEVICE_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_NONE,
+				      error)) {
+		g_prefix_error(error, "write row command error: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_pure_hid_device_write_firmware(FuDevice *device,
+				       FuFirmware *firmware,
+				       FuProgress *progress,
+				       FwupdInstallFlags flags,
+				       GError **error)
+{
+	FuCcgxPureHidDevice *self = FU_CCGX_PURE_HID_DEVICE(device);
+	guint8 fw_mode = 1;
+	GPtrArray *records = fu_ccgx_firmware_get_records(FU_CCGX_FIRMWARE(firmware));
+
+	g_debug("operating mode: 0x%02x", self->operating_mode);
+
+	if (self->operating_mode != FU_CCGX_PURE_HID_FW_MODE_FW2) {
+		fw_mode = 2;
+	}
+	fu_progress_set_status(progress, FWUPD_STATUS_DEVICE_WRITE);
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, records->len);
+
+	for (guint i = 0; i < records->len; i++) {
+		FuCcgxFirmwareRecord *record = g_ptr_array_index(records, i);
+		g_debug("writing row 0x%x 0x%04lx",
+			record->row_number,
+			g_bytes_get_size(record->data));
+		if (!fu_ccgx_pure_hid_write_row(self,
+						record->row_number,
+						g_bytes_get_data(record->data, NULL),
+						g_bytes_get_size(record->data),
+						error))
+			return FALSE;
+		fu_progress_step_done(progress);
+	}
+
+	g_debug("bootswitch");
+	if (!fu_ccgx_pure_hid_command(self, FU_CCGX_PURE_HID_COMMAND_SET_BOOT, fw_mode, error)) {
+		g_prefix_error(error, "bootswitch command error: ");
+		return FALSE;
+	}
+
+	g_debug("reset");
+	if (!fu_ccgx_pure_hid_command(self,
+				      FU_CCGX_PURE_HID_COMMAND_JUMP,
+				      FU_CCGX_PD_RESP_DEVICE_RESET_CMD_SIG,
+				      error)) {
+		g_prefix_error(error, "reset command error: ");
+		return FALSE;
+	}
+
+	g_debug("wait for replug");
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+
```

Comment:
all these `g_debug` lines probably can go, they're more bringup related than valuable to debug IMO